Apply action to all selected files in solve merge conflict

racecrew opened the pull request:
<!-- Please read CONTRIBUTING.md before submitting a pull request -->

Fixes #6279

## Proposed changes
Add a checkbox to apply an action to all conflicted files which an user has selected and wants to get handled at once without promting each time a dialog

## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

![SolveMergeConflict_Dialog_Before](https://user-images.githubusercontent.com/17234283/94859594-11012e80-0435-11eb-8e7b-b8b6e5ee6ea0.png)


### After

![SolveMergeConflict_Dialog_After](https://user-images.githubusercontent.com/17234283/94859622-1cecf080-0435-11eb-8be5-bb23c25edbff.png)

## Test methodology <!-- How did you ensure quality? -->

Manual

## Test environment(s) <!-- Remove any that don't apply -->

- GIT <!-- Add version 2.11 or above -->
- Windows <!-- Add version 7 SP1 or above -->

<!-- Mention language, UI scaling, or anything else that might be relevant -->


----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

gerhardol reviewed (commented):
checkbox must be above buttons

RussKie commented:
> checkbox must be above buttons

Actually I am going to suggest the keep it but use a Task Dialog instead, like we have for the push
![image](https://user-images.githubusercontent.com/4403806/94877264-5e18eb00-049d-11eb-8dfa-c1a0c457c436.png)

RussKie reviewed (changes requested):
Let's take a little further and delete `FormModifiedDeletedCreated` replacing it with a Task Dialog (have a look at https://github.com/gitextensions/gitextensions/pull/8291).

Thank you

RussKie commented on GitUI/CommandsDialogs/FormResolveConflicts.cs:
```diff
@@ -53,6 +53,8 @@ public partial class FormResolveConflicts : GitModuleForm
         private readonly TranslationString _keepModifiedButtonText = new TranslationString("Keep modified");
         private readonly TranslationString _keepBaseButtonText = new TranslationString("Keep base file");
 
+        private readonly TranslationString _applyToAllCheckBoxText = new TranslationString("Apply to all");
```
It would be a better UX to say "this action applies to 2/3/10/... files" as opposed to "all" files. Will it apply to _all_ files or _all selected_ files?
Also the checkbox shouldn't be shown if there is only one file is selected.

racecrew replied:
Yes, you are completely right.

RussKie commented:
Also please provide your sign-off in contributors.txt. Thanks

racecrew commented on GitUI/CommandsDialogs/FormResolveConflicts.cs:
```diff
@@ -95,6 +97,9 @@ public partial class FormResolveConflicts : GitModuleForm
 
         private readonly IFullPathResolver _fullPathResolver;
 
+        private bool _ResolveConflictsApplyToAll = false;
+        private string _ResolveConflictsApplyToAllLastAction = "";
```
I'll change this

RussKie commented:
Please update all screenshots

On Sat, 3 Oct. 2020, 8:14 am Otis Hix, <notifications@github.com> wrote:

> @racecrew <https://github.com/racecrew> requested your review on: #8501
> <https://github.com/gitextensions/gitextensions/pull/8501> Apply action
> to all selected files in solve merge conflict.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/8501#event-3835808594>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXQ3MMY7ONOONPUWQELSIZGCVANCNFSM4SA4ZTYA>
> .
>

racecrew commented:
**One file is selected**

![SolveMergeConflict_TaskDialog_OneFile_Selected](https://user-images.githubusercontent.com/17234283/94991704-0bb2f980-0585-11eb-9eea-87d909bc9f7c.png)

**Multiple files are selected**

![SolveMergeConflict_TaskDialog_MultipleFilesSelected](https://user-images.githubusercontent.com/17234283/94991779-7fed9d00-0585-11eb-957c-bb5f5be8eb4d.png)


By selecting multiple files some of them could have a "different type of change". The conflicted files are grouped as:
* deleted locally and modified remotely 
* modified locally and deleted remotely
* remained files

https://github.com/gitextensions/gitextensions/blob/c411ac64c5946ebdedf669f81911543da5dbe804/GitUI/CommandsDialogs/FormResolveConflicts.cs#L1209-L1229

The task dialog will prompt for each group and "apply to all" is remembered for only one group. Before processing a new conflict group the dialog will prompt. For the "remained files" the checkbox "apply to all" is hidden. For each of this file the dialog will prompt.

**deleted locally and modified remotely**

![SolveMergeConflict_TaskDialog_MultipleFilesSelected_DeletedLocallyAndModifiedRemotely](https://user-images.githubusercontent.com/17234283/94992143-3d798f80-0588-11eb-915c-97355e993dd1.png)

**modified locally and deleted remotely**

![SolveMergeConflict_TaskDialog_MultipleFilesSelected_ModifiedLocallyAndDeletedRemotely](https://user-images.githubusercontent.com/17234283/94992068-9b59a780-0587-11eb-96a5-e7eff69788e7.png)

RussKie commented on GitUI/GitUI.csproj:
```diff
@@ -66,10 +66,13 @@
   </ItemGroup>
 
   <ItemGroup>
+    <Compile Remove="CommandsDialogs\ResolveConflictsDialog\**" />
     <Compile Remove="Resources\Icons\artwork\**" />
     <Compile Remove="Resources\Icons\originals\**" />
+    <EmbeddedResource Remove="CommandsDialogs\ResolveConflictsDialog\**" />
     <EmbeddedResource Remove="Resources\Icons\artwork\**" />
     <EmbeddedResource Remove="Resources\Icons\originals\**" />
+    <None Remove="CommandsDialogs\ResolveConflictsDialog\**" />
```
I don't believe any of these are needed

RussKie commented on GitUI/CommandsDialogs/FormResolveConflicts.cs:
```diff
@@ -94,6 +99,21 @@ public partial class FormResolveConflicts : GitModuleForm
+        }
+
+        private SolveMergeConflictActionType _solveMergeConflictDialogResult;
+        private bool _SolveMergeConflictApplyToAll;
```
Here and below, fields must follow the _camelCase_ notation
```suggestion
        private bool _solveMergeConflictApplyToAll;
```

RussKie commented on GitUI/CommandsDialogs/FormResolveConflicts.cs:
```diff
@@ -94,6 +99,21 @@ public partial class FormResolveConflicts : GitModuleForm
+        private bool _SolveMergeConflictApplyToAll;
+        private SolveMergeConflictActionType _SolveMergeConflictLastAction;
+        private string _SolveMergeConflictDialogCheckboxText;
+        private delegate void SolveMergeConflictCheckRevisionDelegate(SolveMergeConflictActionType dialogResult);
```
Not needed, use `Action<SolveMergeConflictActionType>` instead

RussKie commented on GitUI/CommandsDialogs/FormResolveConflicts.cs:
```diff
@@ -94,6 +99,21 @@ public partial class FormResolveConflicts : GitModuleForm
         #endregion
 
         private readonly IFullPathResolver _fullPathResolver;
+        private enum SolveMergeConflictActionType
+        {
+            atNone = 0,
+            atKeepLocal = 1,
+            atKeepRemote = 2,
+            atKeepBase = 3
+        }
```
```suggestion
        private enum ConflictResolutionPreference
        {
            None = 0,
            KeepLocal = 1,
            KeepRemote = 2,
            KeepBase = 3
        }
```

RussKie commented on GitUI/CommandsDialogs/FormResolveConflicts.cs:
```diff
@@ -874,38 +894,115 @@ private void ChooseRemoteOnConflict(string fileName)
+            string dialogText, string dialogInstructionText, string dialogCaption, string dialogFooterCheckboxText,
+            string keepLocalButtonText, string keepRemoteButtonText, string keepBaseButtonText)
+        {
+            bool shouldOpenDialog = _SolveMergeConflictApplyToAll == false;
+            if (shouldOpenDialog)
```
Non nullable bools don't need this notation. The same applies below to other bool operations.

The temp variable is not required:
```suggestion
            if (!_solveMergeConflictApplyToAll)
```

RussKie commented on GitUI/CommandsDialogs/FormResolveConflicts.cs:
```diff
@@ -874,38 +894,115 @@ private void ChooseRemoteOnConflict(string fileName)
+        {
+            SolveMergeConflictCheckRevisionDelegate executeAction = (dialogResult) =>
+            {
+                switch (dialogResult)
```
Use pattern matching, it will make the code denser and easier to read

racecrew replied:
Honestly speaking, I am not sure what you are meaning with that. I did change the code, maybe the change is somehow already there ;-) Please take a look, thank you!

RussKie commented:
> **Multiple files are selected**
> 
> ![SolveMergeConflict_TaskDialog_MultipleFilesSelected](https://user-images.githubusercontent.com/17234283/94991779-7fed9d00-0585-11eb-957c-bb5f5be8eb4d.png)
> 

This is totally non obvious, and instead of 3 it should say 5. Or the text has to clearly explain why it is 3.

RussKie commented:
Thank you for the screenshots 🚀

racecrew commented:
> > **Multiple files are selected**
> > ![SolveMergeConflict_TaskDialog_MultipleFilesSelected](https://user-images.githubusercontent.com/17234283/94991779-7fed9d00-0585-11eb-957c-bb5f5be8eb4d.png)
> 
> This is totally non obvious, and instead of 3 it should say 5. ~Or the text has to clearly explain why it is 3.~
> 
> [EDIT] Apparently I didn't read the top text :) But the fact that the selection contains 5 entries and the bottom checkbox has 3 is head-spinning...
> I think it should say something like "Apply to all selected". I know it kind of goes back to the original wording, but I didn't realise there could be grouping like the top message of the dialogs, which has numbers that may be different to selection count.

I have reworked the text messages, here are the new screenshots... I hope, this time it looks better, I did my best :-)

**Selected files of the same group, e.g. "deleted locally and modified remotely"**

![1_SelectedFilesOfSameGroup](https://user-images.githubusercontent.com/17234283/95003356-cde2bf00-05de-11eb-8b61-f36617a4a10c.png)

**Selected files of the different groups, e.g. "deleted locally and modified remotely" and vice versa"**

![2_SelectedFilesOfDifferentGroups_DeletedLocaly](https://user-images.githubusercontent.com/17234283/95003380-04b8d500-05df-11eb-8a3a-cb4c20bf11ee.png)

![3_SelectedFilesOfDifferentGroups_DeletedRemotely](https://user-images.githubusercontent.com/17234283/95003381-07b3c580-05df-11eb-9abb-340ca6133e7b.png)

racecrew commented:
In addition, if the user doesn't use the "apply" checkbox, the counter of the files is decremented, e.g.

![1_SelectedFilesOfSameGroup](https://user-images.githubusercontent.com/17234283/95003449-b6f09c80-05df-11eb-8f00-f705b21d9d7d.png)

After an action the checkbox text will change:
- Apply to f1.txt and 2 other file(s)
- Apply to f2.txt and 1 other file(s)
- Checkbox will be hidden because one file will be left

RussKie commented:
Thank you.

I'm still struggling with visual disconnect between selections and words... But the whole conflict resolution subject is head spinning :)
So I suppose this may work (with few more tweaks, please read on).




> **Selected files of the different groups, e.g. "deleted locally and modified remotely" and vice versa"**
> 
> ![2_SelectedFilesOfDifferentGroups_DeletedLocaly](https://user-images.githubusercontent.com/17234283/95003380-04b8d500-05df-11eb-8a3a-cb4c20bf11ee.png)

I personally think the bolded text in "_f4.txt and 2 other **of 5** selected files..._" in unnecessary.

So what happens to other 3 files if a user checks the box and proceeds to deleting? The message doesn't specify this, so if I were the user I'd probably cancel and then selected files invidiually.

> In addition, if the user doesn't use the "apply" checkbox, the counter of the files is decremente

👍

racecrew commented:
> Thank you.
> 
> I'm still struggling with visual disconnect between selections and words... But the whole conflict resolution subject is head spinning :)
> So I suppose this may work (with few more tweaks, please read on).
> 
> > **Selected files of the different groups, e.g. "deleted locally and modified remotely" and vice versa"**
> > ![2_SelectedFilesOfDifferentGroups_DeletedLocaly](https://user-images.githubusercontent.com/17234283/95003380-04b8d500-05df-11eb-8a3a-cb4c20bf11ee.png)
> 
> I personally think the bolded text in "_f4.txt and 2 other **of 5** selected files..._" in unnecessary.
> 
> So what happens to other 3 files if a user checks the box and proceeds to deleting? The message doesn't specify this, so if I were the user I'd probably cancel and then selected files invidiually.
> 
> > In addition, if the user doesn't use the "apply" checkbox, the counter of the files is decremente
> 
> 👍

The other 3 files are belonging to "other group", in this case they are _deleted remotely and modified locally_. So, after the user applied the same action to f1.txt and f3.txt, the task dialog will appear again with the msg _"f4.txt and 2 other of 5 selected..."_

So, the Task Dialog will prompt for each of such groups.... 

You are right, the whole topic is not easy to handle :-)

racecrew commented:
> The message doesn't specify this, so if I were the user I'd probably cancel and then selected files invidiually.

The user can do this, it would be nothing wrong, but if the user would check the box, then a task dialog will prompt for the new group of selected items.

gerhardol reviewed (commented):
Seem OK, only browsed the code
Wordings seem sensible to me (I would use "out of" instead of just "of".

The resolution form need to associate the "user branch" and possibly "target" branch with local/remote, ours/theirs, but that is a separate task (not so easy as it seems).

RussKie commented:
> Seem OK, only browsed the code
> Wordings seem sensible to me (I would use "out of" instead of just "of".

Alright, let's fix the wording and take it.

Please resolve the MC as well. Thanks

RussKie commented on GitUI/CommandsDialogs/FormResolveConflicts.cs:
```diff
@@ -1053,16 +1224,75 @@ private void OpenMergeTool()
+                    }
+
+                    // This is needed the task dialog shows the same order of files as selected in the grid
+                    filesDeletedLocallyAndModifiedRemotely.Reverse();
```
This isn't perf optimal - we create a list, add items and the reverse. Why not just add items in the order we need?
I.e. 
```cs
List<ConflictData> filesDeletedLocallyAndModifiedRemotely = new List<ConflictData>();
foreach (var conflictData in items)
{
    filesDeletedLocallyAndModifiedRemotely.Insert(0, conflictData);
}
```

RussKie commented on GitUI/CommandsDialogs/FormResolveConflicts.cs:
```diff
@@ -1007,41 +1148,71 @@ private bool CheckForRemoteRevision(ConflictData item)
+            string dialogCaption = _solveMergeConflictDialogCaption.Text;
+            string dialogFooterCheckboxText = (_filesModifiedLocallyAndDeletedRemotelySolved > 1) ? string.Format(_solveMergeConflictApplyToAllCheckBoxText.Text, item.Filename, _filesModifiedLocallyAndDeletedRemotelySolved - 1) : string.Empty;
+
+            string keepLocalButtonText = string.Format(_keepModifiedButtonText.Text + " ({0})", GetLocalSideString());
```
This (and similar) is done in a number of places for seemingly no good reason - we're creating many unnecessary string objects.
```suggestion
            string keepLocalButtonText = string.Concat(_keepModifiedButtonText.Text, $" (GetLocalSideString())");
```

racecrew replied:
Okay, I will remove local string objects and pass the values directly as arguments

RussKie commented on GitUI/CommandsDialogs/FormResolveConflicts.cs:
```diff
@@ -94,6 +101,25 @@ public partial class FormResolveConflicts : GitModuleForm
         #endregion
 
         private readonly IFullPathResolver _fullPathResolver;
+        private enum ConflictResolutionPreference
+        {
+            None = 0,
+            KeepLocal = 1,
+            KeepRemote = 2,
+            KeepBase = 3
+        }
+
```
```suggestion
        private enum ConflictResolutionPreference
        {
            None = 0,
            KeepLocal = 1,
            KeepRemote = 2,
            KeepBase = 3
        }
        
        private readonly IFullPathResolver _fullPathResolver;
```

gerhardol commented:
Merge is evil...
I rebased your changes and squashed: https://github.com/gerhardol/gitextensions/tree/feature/ApplyActionToAllSelectedFilesInSolveMergeConflict

You can either just rebase interactively on master, change all but the first pick to s (squash) and merge the commit message
Or just reset hard to my commit

gerhardol reviewed (commented):
Briefly reviewed, has not run
Seem OK to me
(@RussKie will never believe that I give a lot of comments at work reviews!)

racecrew commented on GitUI/CommandsDialogs/FormResolveConflicts.cs:
```diff
@@ -1053,16 +1184,72 @@ private void OpenMergeTool()
+                        ResolveItemConflict(conflictData);
+                    }
+
+                    // For the remained files it should be asked for each file separately which one should be taken
```
It means, for each file the task dialog asks user what to do

racecrew commented on GitUI/CommandsDialogs/FormResolveConflicts.cs:
```diff
@@ -1053,16 +1184,72 @@ private void OpenMergeTool()
+
+                    List<ConflictData> filesDeletedLocallyAndModifiedRemotely = new List<ConflictData>();
+                    List<ConflictData> filesModifiedLocallyAndDeletedRemotely = new List<ConflictData>();
+                    List<ConflictData> filesRemained = new List<ConflictData>();
```
Okay, thanks!

racecrew commented:
How do I resolve the conflict in **contributors.txt**?!

RussKie commented:
> How do I resolve the conflict in **contributors.txt**?!

Rebase :)

RussKie commented:
> will never believe that I give a lot of comments at work reviews!

Hahahah! Don't tell me you ask for tests at work too.

racecrew commented:
> > How do I resolve the conflict in **contributors.txt**?!
> 
> Rebase :)

Again?! :-O

RussKie commented:
As a developer one constantly fetches and rebases...
Only the first few times rebase feels difficult, then it becomes almost
mechanical.

On Wed, 7 Oct. 2020, 4:38 pm Otis Hix, <notifications@github.com> wrote:

> How do I resolve the conflict in *contributors.txt*?!
>
> Rebase :)
>
> Again?! :-O
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/8501#issuecomment-704705013>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXQHGMQDJHETPWRR3QDSJP5G5ANCNFSM4SA4ZTYA>
> .
>

gerhardol commented:
There is a reason why contributors.txt should be a separate pr...

OT: I prefer fast forward merges, not rewriting commits at merges. This is especially _needed_ to work with sub modules But github has no support for ff-only.

gerhardol commented:
Dont merge, rebase...

racecrew commented:
> Dont merge, rebase...

I am struggling here.... I have merged, fetched, sqaushed into one commit.. still getting the conflict... :-(

Meanwhile I know that I should rebase

racecrew commented:
Yeahaaa!! Finally! Thank you guys for your help!!

RussKie commented:
Thank you
